Add warning support to inline cpp_extensions

albanD opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #27154 Handle warning in torchscript
* **#27153 Add warning support to inline cpp_extensions**
* #26614 add pybind version of HANDLE_TH_ERRORS
* #26613 Buffer python warning to avoid deadlocks

Differential Revision: [D17692506](https://our.internmc.facebook.com/intern/diff/D17692506)

albanD committed:
Add warning support to inline cpp_extensions

[ghstack-poisoned]

albanD committed:
Update on "Add warning support to inline cpp_extensions"


Differential Revision: [D17692506](https://our.internmc.facebook.com/intern/diff/D17692506)

[ghstack-poisoned]

ezyang commented on torch/utils/cpp_extension.py:
```diff
@@ -719,6 +720,10 @@ def load_inline(name,
+        with_pytorch_error_handling: Determines whether pytorch error and
+            warning macros are handled by pytorch instead of pybind. This
+            requires an intermediary function and should be deactivated
+            if this intermediary function cause issues.
```
I'm not sure, as a user, what I'm supposed to do with this information. What does an "intermediary function" do? How can I tell if it's causing issues? If we want to let users debug problems, it would be better to tell them all the gory details of how our binding code is generated, so they can make an informed decision here.

ezyang commented on torch/utils/cpp_extension.py:
```diff
@@ -741,11 +746,20 @@ def load_inline(name,
+        cpp_sources.append('template <typename ...Args>')
+        cpp_sources.append('auto _safe_{0}(Args ...args) -> decltype({0}(args...)) {{'.format(function_name))
+        cpp_sources.append('HANDLE_TH_ERRORS')
+        cpp_sources.append('return {0}(args...);'.format(function_name))
```
I think you should use perfect forwarding here. So `Args&& ...` and then `std::forward<Args>(args...)` (or something like that)

ezyang commented on torch/utils/cpp_extension.py:
```diff
@@ -741,11 +746,20 @@ def load_inline(name,
 
     cpp_sources.insert(0, '#include <torch/extension.h>')
 
+    def add_safe_version(function_name):
```
I think, for future readers, it will be helpful if there's a little comment saying what safe versions do.

ezyang commented:
Implementation looks good. Just some comments about docs.

albanD committed:
Update on "Add warning support to inline cpp_extensions"


Differential Revision: [D17692506](https://our.internmc.facebook.com/intern/diff/D17692506)

[ghstack-poisoned]

albanD commented:
@ezyang bad use of rebase/ghstack on my side... This got merged into https://github.com/pytorch/pytorch/pull/26614 so closing here.
